### ISSUE TITLE
Merge into ocds-babel instead of having two libraries?

### ISSUE THREAD

**Issue by jpmckinney** [openownership/bods-babel#4]
I think the main things to change in ocds-babel would be:

* Allow configuration of header constants
* Change if-statements in `translate` a bit
* Make the Sphinx dependency optional

Besides those differences, it looks like bods-babel only removes functions that are unique to OCDS, and renames the Babel extractors.

I prefer having a common library if possible, to avoid repeating (and likely forgetting to repeat) improvements / bugfixes across repositories. What do you think?

**Comment by kd-ods** [comment]
@jpmckinney - will ask @rhiaro to look at this (since she set up the repo off the back of looking at the OCDS setup) and reply.

**Comment by rhiaro** [comment]
It makes sense to me to keep them separate, given BODS doesn't need the OCDS specific parts and it's not impossible that there may be BODS specific additions as the standard develops.

If they were in one library, with the potential to add extractors for other standards in future, as well as the refactoring mentioned we'd need to rename it and move it so it isn't under the remit of a standard-specific organisation I think?

**Comment by jpmckinney** [comment]
Yes, we're open to that (it won't require much work).

To address all expressed concerns, we can have each of bods-babel and ocds-babel use a common dependency (extracted from ocds-babel into some standard-independent org and repo).

**Comment by ghost** [comment]
Is it ok if we continue with this for now but keep this issue open as a possibility for later? I'm mainly drawing on the rule-of-three as in "a reusable component should be tried out in three different applications before it will be sufficiently general to accept into a reuse library. " - https://blog.codinghorror.com/rule-of-three/ has a good writeup.

I also generally agree with the point in the blog about it being more work to split things into reusable libraries, so we would have to be careful weighting up the priority of this.

**Comment by jpmckinney** [comment]
This isn't a rule-of-three scenario. bods-babel is essentially a fork of ocds-babel with only the small changes described in the issue description. This scenario is "should we fork, or should we change upstream?" From what I can see, it likely took more time to delete unused code, rename modules, and update tests accordingly, than it would have taken to change upstream…

If I make those changes upstream, then BODS will be able use ocds-babel without any issue. If at some future date BODS needs something very different, then that's a good time to weigh fork vs. reuse.

The [rule-of-three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) that I'm familiar with is about similar code (written independently) and about methods. It isn't about entire libraries that are copy-pasted (identical code). In any case, ocds-babel is itself already a refactoring…  Realistically, there will also never be a third fork for a third standard.

**Update:** I was familiar with the rule of three in the context of refactoring. In the blog post, it's more generally quoted as "b. a reusable component should be tried out in three different applications before it will be sufficiently general to accept into a reuse library."

In other words, the rule can be applied to use the same library in three contexts, and make changes to the library to suit those contexts. That's pretty much what I'm suggesting we try to do.

**Comment by jpmckinney** [comment]
Anyway, taking a step back from technical details, my hope is that standards can start sharing more software and other resources, and this seems like a very easy one to try that with. It's a very small library that does something low-level, and that presently has no anticipated future work.

In short, would you be open to sharing +code+, instead of forking? If so, I can start on the changes in ocds-babel. OO and OCP can then decide where the library should live and what to name it.

**Comment by ghost** [comment]
> my hope is that standards can start sharing more software and other resources

Taking a step back, I do agree that is a good goal.

> In short, would you be open to sharing

I'm not totally sure what you mean by sharing; but we are happy to try something tho. If you are happy to start working on the changes to ocds-babel, we can try using that directly for a bit. I notice ocds-babel is already versioned with a changelog; this will help, thanks! We can then switch over to using that and deprecate this repository, and if that looks like it will work we can discuss a suitable name and location for this library.

(One thing is that we wouldn't be able to look at switching to it this sprint; that is already fully specced and there are some quite big bits of work with deadlines coming up for BODS that are taking our focus. But we plan no further development on this repo in this sprint and we can look soon.)

**Comment by jpmckinney** [comment]
Great! I ran:

    diff --exclude=.git -ruw path/to/bods-babel/ocds_babel path/to/bods-babel/bods_babel | less

And the differences were as noted earlier.

It looks like https://github.com/openownership/data-standard/pull/188 took advantage of the pattern [described here](https://ocds-standard-development-handbook.readthedocs.io/en/latest/standard/translation/implementation/) and [implemented here](https://github.com/open-contracting/standard_profile_template), so using the new version of ocds-babel (or whatever it's called) should only require:

* renaming the module imported in conf.py
* adding an argument to the `translate` call
* renaming the entry points used in `*.cfg` files
* adding a `headers = title,description,technical note` option to `babel_bods_codelist.cfg`

See ocds-babel's [changelog](https://github.com/open-contracting/ocds-babel/blob/master/CHANGELOG.md) and [readme](https://github.com/open-contracting/ocds-babel#readme) (once PR merged). PR is https://github.com/open-contracting/ocds-babel/pull/8 and will be merged once Travis runs.

**Comment by jpmckinney** [comment]
PR is merged and I've updated all repos that use ocds-babel.

**Comment by Bjwebb** [comment]
It looks to me like this should be a quick change to make, and we can check all the output for any changes. I'd start with a 2 hour timebox on this, and if it takes longer we can reassess.

**Comment by jpmckinney** [comment]
Note the docs relevant to BODS are on this page: https://ocds-babel.readthedocs.io/en/latest/api/extract.html

**Comment by Bjwebb** [comment]
Looks like we can make this change without any changes to translation output, PR at https://github.com/openownership/data-standard/pull/291

**Comment by Bjwebb** [comment]
PR merged.
